src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -0,0 +1,186 @@
+import com.intellij.openapi.project.Project;
+import fi.aalto.cs.apluscourses.intellij.notifications.MissingFileNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.NetworkErrorNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.NotSubmittableNotification;
+import fi.aalto.cs.apluscourses.intellij.notifications.Notifier;
+import fi.aalto.cs.apluscourses.intellij.services.MainViewModelProvider;
+import fi.aalto.cs.apluscourses.intellij.services.PluginSettings;
+import fi.aalto.cs.apluscourses.intellij.utils.VfsUtil;
+import fi.aalto.cs.apluscourses.model.APlusAuthentication;
+import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.Exercise;
+import fi.aalto.cs.apluscourses.model.Group;
+import fi.aalto.cs.apluscourses.model.SubmissionHistory;
+import fi.aalto.cs.apluscourses.model.SubmittableExercise;
+import fi.aalto.cs.apluscourses.presentation.APlusAuthenticationViewModel;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+import fi.aalto.cs.apluscourses.presentation.MainViewModel;
+import fi.aalto.cs.apluscourses.presentation.exercise.ExerciseViewModel;
+import fi.aalto.cs.apluscourses.presentation.exercise.ExercisesTreeViewModel;
+import fi.aalto.cs.apluscourses.presentation.exercise.SubmissionViewModel;
+import fi.aalto.cs.apluscourses.ui.exercise.ModuleSelectionDialog;
+import fi.aalto.cs.apluscourses.ui.exercise.SubmissionDialog;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import org.jetbrains.annotations.CalledWithReadLock;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class SubmitExerciseAction extends AnAction {
+
+  public static final String ACTION_ID = SubmitExerciseAction.class.getCanonicalName();
+
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull
+  private Notifier notifier;
+
+  public SubmitExerciseAction() {
+    this(PluginSettings.getInstance(), Notifications.Bus::notify);
+  }
+
+  public SubmitExerciseAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                              @NotNull Notifier notifier) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.notifier = notifier;
+  }
+
+  @Override
+  public void update(@NotNull AnActionEvent e) {
+    MainViewModel mainViewModel = mainViewModelProvider.getMainViewModel(e.getProject());
+    CourseViewModel courseViewModel = mainViewModel.courseViewModel.get();
+    ExercisesTreeViewModel exercisesTreeViewModel
+        = mainViewModel.exercisesViewModel.get();
+    APlusAuthenticationViewModel authenticationViewModel
+        = mainViewModel.authenticationViewModel.get();
+
+    e.getPresentation().setEnabled(e.getProject() != null && exercisesTreeViewModel != null
+        && authenticationViewModel != null && courseViewModel != null
+        && courseViewModel.getModel() != null);
+  }
```

Comment:
Nit: This `update()` method has quite a lot logic in it plus it accesses several synchronized variables. Perhaps a better solution would be moving this logic to the main view-model (say `isSubmitAvailable()` method). To (micro-?)optimize it, the value should be stored in a (volatile) boolean field which should be updated when one of the `courseViewModel`, `exercisesViewModel`, etc. changes (in the constructor `MainViewModel`, add `courseViewModel.addValueObserver(dummy -> updateIsSubmitAvailable())` and similar). Maybe even easier solution is to make `isSubmitAvailable` yet another observable (read-only) property in the main view-model. In this case, add `isSubmitAvailable.declareDependentOn(courseViewModel)` and similar to the constructor.

Reply:
Good point! :+1: I'll think about which of the alternatives I'll go with